src/ddb/ddb_vos.c

```diff
@@ -0,0 +1,584 @@
+		return handle_sv(cb_arg, entry);
+	case VOS_ITER_RECX:
+		return handle_array(cb_arg, entry);
+	case VOS_ITER_DTX:
+		printf("dtx\n");
+		break;
+	case VOS_ITER_NONE:
+		break;
+	}
+
+	return 0;
+}
+
+static int
+iter_cont_recurse_cb(daos_handle_t ih, vos_iter_entry_t *entry, vos_iter_type_t type,
+		     vos_iter_param_t *param, void *cb_arg, unsigned int *acts)
+{
+	vos_iter_param_t	 cont_param = {0};
+	struct vos_iter_anchors	 anchors = {0};
+	daos_handle_t		 coh;
+	int			 rc;
+
+	D_ASSERT(type == VOS_ITER_COUUID);
+
+	rc = handle_cont(cb_arg, entry, param);
+	if (!SUCCESS(rc))
+		return rc;
+
+	/* recursively iterate the objects in the container */
+	rc = vos_cont_open(param->ip_hdl, entry->ie_couuid, &coh);
+	if (!SUCCESS(rc))
+		return rc;
+
+	cont_param.ip_hdl = coh;
+	cont_param.ip_epr.epr_hi = DAOS_EPOCH_MAX;
+	rc = ddb_vos_iterate(&cont_param, VOS_ITER_OBJ, true, &anchors, handle_iter_cb, cb_arg);
+
+	if (rc != 0)
+		D_ERROR("vos_iterate error: "DF_RC"\n", DP_RC(rc));
+
+	rc = vos_cont_close(coh);
+
+	return rc;
+}
+
+static int
+iter_cont_recurse(vos_iter_param_t *param, struct ddb_iter_ctx *ctx)
+{
+	struct vos_iter_anchors	anchors = {0};
+
+	return ddb_vos_iterate(param, VOS_ITER_COUUID, false, &anchors, iter_cont_recurse_cb, ctx);
+}
+
+int
+dv_iterate(daos_handle_t poh, struct dv_tree_path *path, bool recursive,
+	   struct vos_tree_handlers *handlers, void *handler_args)
+{
+	vos_iter_param_t	param = {0};
+	struct vos_iter_anchors	anchors = {0};
+	int			rc;
+	daos_handle_t		coh = DAOS_HDL_INVAL;
+	vos_iter_type_t		type;
+	struct ddb_iter_ctx	ctx = {0};
+
+	ctx.handlers = handlers;
+	ctx.handler_args = handler_args;
+	ctx.poh = poh;
+
+	param.ip_epr.epr_hi = DAOS_EPOCH_MAX;
+
+	if (uuid_is_null(path->vtp_cont)) {
+		param.ip_hdl = poh;
+
+		if (recursive)
+			/*
+			 * currently vos_iterate doesn't handle recursive iteration starting with a
+			 * container. This works around that limitation.
+			 */
+			return iter_cont_recurse(&param, &ctx);
+		return ddb_vos_iterate(&param, VOS_ITER_COUUID, false, &anchors,
+				       handle_iter_cb, &ctx);
+	}
+
+	rc = vos_cont_open(poh, path->vtp_cont, &coh);
+	if (!SUCCESS(rc))
+		return rc;
+
+	param.ip_hdl = coh;
+
+	param.ip_oid = path->vtp_oid;
+	param.ip_dkey = path->vtp_dkey;
+	param.ip_akey = path->vtp_akey;
+
+	if (daos_oid_is_null(path->vtp_oid.id_pub)) {
+		type = VOS_ITER_OBJ;
+	} else if (path->vtp_dkey.iov_len == 0) {
+		type = VOS_ITER_DKEY;
+	} else if (path->vtp_akey.iov_len == 0) {
+		type = VOS_ITER_AKEY;
+	} else {
+		/* Don't know if the akey value is sv or array so just
+		 * try to iterate both. Doesn't seem to have any negative consequences for
+		 * trying to iterate what the value is not.
+		 */
+		rc = ddb_vos_iterate(&param, VOS_ITER_RECX, recursive, &anchors,
+				     handle_iter_cb, &ctx);
+		if (!SUCCESS(rc)) {
+			vos_cont_close(coh);
+
+			return rc;
+		}
+		rc = ddb_vos_iterate(&param, VOS_ITER_SINGLE, recursive, &anchors,
+				     handle_iter_cb, &ctx);
```

Comment:
Jeff and I talked about how to make this better. I'd like to leave it like this for now and resolve in my next patch.